scripts/Muon/GUI/Common/thread_model.py

```diff
@@ -78,17 +120,29 @@ def loadData(self, inputs):
                                  " loadData() method, which it is attempting to call.")
         self.model.loadData(inputs)
 
-    def threadWrapperSetUp(self, startSlot, endSlot):
-        assert hasattr(startSlot, '__call__')
-        assert hasattr(endSlot, '__call__')
-        self.start_slot, self.end_slot = startSlot, endSlot
-        self.started.connect(self.start_slot)
-        self.finished.connect(self.end_slot)
-        self.finished.connect(self.threadWrapperTearDown)
+    def threadWrapperSetUp(self,
+                           on_thread_start_callback=lambda: 0,
+                           on_thread_end_callback=lambda: 0,
+                           on_thread_exception_callback=None):
```

Comment:
why does `on_thread_exception_callback` get set to `None` and the others are `lambda`

Reply:
You can't (so far as I am aware) reference `self` inside the argument list. I want to be able to use `self._exception_callback` which holds the default callback (which is the function that was called before the changes were made to keep consistency) as the default argument to this function.

Hence, I have to set it to "None" by default, then do an `is None` check inside the function body.

Reply:
Regarding the tear down; the error handling happens through the `error` signal, connected to the `warning` slot. This is connected and disconnected correctly in the setup/teardown; the `warning` slot itself calls the `self._exception_callback` member variable; hence the `setUp` function only needs to assigned the `on_thread_exception_callback` to the `self._exception_callback` and not worry about connect/disconnect statements.

Reply:
1. So why are the other two lambdas and not None? 
2. Looking at your code, am I right in saying that you create a thread object and keep the same object setting it running each time it is needed?

Reply:
1. The other two just don't need to be `None`, as this would incur the following code in the function body;

```
if on_thread_start_callback is not None:
	self.start_slot = lambda: 0
```

and similarly for `on_thread_end_callback`

and it is easier to just declare the default in the argument list.

2. If I understand correctly, yes. You can either create new instances of `ThreadModel` as and when you need them (as before); or you can simply create one instance and hit `.start()` whenever you need to run it. I feel the latter option is slightly more readable in code but both would work fine.